Adding recipes.replaceAllOccurences(IIngredient,IIngredient) as a new zenscript command

ZenDarva opened the pull request:
The command will search through all vanilla style recipes for any recipe containing the first ingredient, remove that recipe, and create a new recipe that is identical, except the second ingredient has been substituted for the first.

I feel this will be useful in cases such as replacing all Machine Frames from one mod to another, or (the case that made me write this) switching all Industrial Foregoing recipes to use Pneumaticraft Plastic instead of IF Plastic.

Please let me know if you simply don't like the idea, or see anything you would want changed before being able to accept this pull request, and I'll get right on it.

ZenDarva committed:
Adding recipes.replaceAllOccurences(IIngredient,IIngredient) as a new zenscript command.

It will search through all vanilla style recipes for any recipe containing the first ingredient, remove that recipe, and create a new recipe that is identical, except the second ingredient has been substituted for the first.

I feel this will be useful in cases such as replacing all Machine Frames from one mod to another, or (the case that made me write this) switching all Industrial Foregoing recipes to use Pneumaticraft Plastic instead of IF Plastic.

jaredlll08 commented:
Any idea how performant this it?

On Tue, Mar 13, 2018, 04:57 ZenDarva <notifications@github.com> wrote:

> The command will search through all vanilla style recipes for any recipe
> containing the first ingredient, remove that recipe, and create a new
> recipe that is identical, except the second ingredient has been substituted
> for the first.
>
> I feel this will be useful in cases such as replacing all Machine Frames
> from one mod to another, or (the case that made me write this) switching
> all Industrial Foregoing recipes to use Pneumaticraft Plastic instead of IF
> Plastic.
>
> Please let me know if you simply don't like the idea, or see anything you
> would want changed before being able to accept this pull request, and I'll
> get right on it.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/CraftTweaker/CraftTweaker/pull/497
> Commit Summary
>
>    - Adding recipes.replaceAllOccurences(IIngredient,IIngredient) as a
>    new zenscript command.
>
> File Changes
>
>    - *M*
>    CraftTweaker2-API/src/main/java/crafttweaker/api/recipes/IRecipeManager.java
>    <https://github.com/CraftTweaker/CraftTweaker/pull/497/files#diff-0>
>    (5)
>    - *M*
>    CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/recipes/MCRecipeManager.java
>    <https://github.com/CraftTweaker/CraftTweaker/pull/497/files#diff-1>
>    (86)
>
> Patch Links:
>
>    - https://github.com/CraftTweaker/CraftTweaker/pull/497.patch
>    - https://github.com/CraftTweaker/CraftTweaker/pull/497.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CraftTweaker/CraftTweaker/pull/497>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AFS9xd-WNJ4ktYJBmE46xVyLOa9_D-GGks5tdzWJgaJpZM4SoAQu>
> .
>

ZenDarva commented:
Quick and dirty profiling says it takes about 76 ms to replace all of the Iron Ingots, in a Dev environment, with no added mods.  That's 34 recipes, so just a bit above 2ms per recipe replaced.  I'll do some more testing with a full modpack when I get home from work.

I'm tempted to write out the longhand zenscript to compare the time taken to do it manually, but avoiding doing that is the specific reason I wrote this.  Heh.  Theoretically, this method should always be faster, simply due to having less overhead.

Two thirds of the processing time is in apply() unsurprisingly.   I'll try and think of any optimizations today while I'm at work, but off the top of my head, I don't see any easy increases.

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/recipes/MCRecipeManager.java:
```diff
@@ -261,7 +267,85 @@ public void removeRecipes(List<ResourceLocation> removingRecipes) {
+            recipes = ForgeRegistries.RECIPES.getEntries();
+
+            for(Map.Entry<ResourceLocation, IRecipe> recipeEntry : recipes) {
+                OreIngredient oi;
```
What's that one for?
I can't see any assignments to it...

ZenDarva replied:
You're absolutely right, this is leftover from debugging, and I simply missed removing it during cleanup, I'll get that fixed tonight.

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/recipes/MCRecipeManager.java:
```diff
@@ -261,7 +267,85 @@ public void removeRecipes(List<ResourceLocation> removingRecipes) {
+    public static class ActionReplaceAllOccurences extends ActionBaseRemoveRecipes {
+        //I'm odd, in that I'm an ActionBaseRemoveRecipes, that also creates recipes.
+        private List<ResourceLocation> toRemove;
+        List<ICraftingRecipe> toChange;
```
If you are only using MCRecipeBase objects, make the list of that type, so you won't need all this casting
Alternatively, there's the possibility of adding those methods to ICraftingRecipe as well

kindlich reviewed (commented):
First look looking good.

Suggestion:
- You could make the recipes use the old names or `oldname_replaced` or something...

ZenDarva committed:
Changed how names are set for replacement recipes
Fixed duplicate detection of recipes in getAllForIngredient
Refactored types to reduce casting.

ZenDarva commented:
Newest update addresses all the issues brought up by kindlich, and fixes a bug in how recipes were detected that was causing duplicate recipes to be registered.

Further performance checking against an Full minecraft instance with 111 mods.  Replacing all instances of minecraft:iron_ingot with ore:plankWood takes 214ms to change 297 recipes.  Oddly, with more items the balance of the time moves to the constructor, with 181ms, whereas apply takes 33ms.

One edge case remains that I haven't been able to detect a reason for.  In Deobf testing, XU2's hopper_shortcut recipe isn't processed correctly.  I cannot reproduce it in the Dev environment, and there doesn't appear to be anything odd about the recipe, so I'm unsure why it's happening.

ZenDarva committed:
Further differentiate recipe names.
Remove leftover testing line.

kindlich commented:
Two Questions:

- Say I replace all occurences of Iron ingots with Oak planks and in the next line all occurences of oak planks with sticks. Would the recipes be changed twice?
- Say I have a recipe with both iron ingots and sticks. If I replace iron ingots with gold ingots and sticks with something else. Would the recipe end up with both changed correctly?

ZenDarva commented:
As the code is written now, the answer to both of those is No, because the adding of the new Recipes happens in a different step than the removing of recipes.  It would have worked in my original approach, because I registered the recipes directly in changeIngredients, rather than creating objects to put into MCRecipeManager.recipesToAdd.  The failing with that approach, was that the recipes did not show up in JEI, because the CraftTweaker JEI plugin uses MCRecipeManager.recipesToAdd as its source of recipes, and that it didn't follow the established conventions of the mod.

In the first case... All the iron ingots would wind up as oak planks, and all the recipes that had oak planks originally would wind up with sticks.

In the second case, You would wind up with Two copies of each recipe, with each change applied separately. 

Obviously, this is not an acceptable case.  To solve it, I'll have to go back to registering the recipes directly in changeIngredients, and create a new class extending ActionBaseAddRecipe, with apply() dummied out, so that JEI can show the recipes correctly.

The second of your two cases should be solved in a slightly different way, with an overloaded version of replaceAllOccurences(), that takes two arrays of IIngredients, and replaces a[0] with b[0], a[1] with b[1] and so on.  I think that's a bit beyond the scope of getting the initial functionality in though.  

Your opinion?

ZenDarva committed:
Added ActionDummyAddRecipe for JEI plugin compatibility.
Moved adding recipes into ActionReplaceAllOccurences, so that subsequent calls to ActionReplaceAllOccurences could modify those results.  This is outside the normal operation of current actions, but is necessary for the code to function correctly.
Moved all work into ActionReplaceAllOccurences apply(), due to timing of changes.
General bug fixes, and speed improvements.

ZenDarva commented:
Ok, updated to fix the case where you use replaceAllOccurences to modify recipes already modified by replaceAllOccurences. 

The recipe adding now occurs during the recipesToRemove apply stage, instead of creating objects that will later add the recipes.  While this doesn't fit the existing paradigm, it's the only reasonable way I could find to do it.

I've tested this against the same instance I've used for previous tests, and all of the recipes I observed appeared to have changed correctly.

Performance metrics:
recipes.replaceAllOccurences(<minecraft:iron_ingot>, <ore:plankWood>);
Removed 379 and added 379 recipes in 417milli
recipes.replaceAllOccurences(<ore:plankWood>, <minecraft:stick>);
Removed 1310 and added 1310 recipes in 1334milli

Run sequentially in the same instance.  So 379 of the recipes replaced in the second command were already modified by the first command.  I did verify that third order replacements work as well, but only in a dev instance, since they use the exact same code as second order replacements.

Crafttweaker log output from a full instance.:
[crafttweaker.log](https://github.com/CraftTweaker/CraftTweaker/files/1840185/crafttweaker.log)

Sorry for the delay, managed to catch the Flu something fierce.
